.pipelines/templates/.builder-release-template-windows.yaml

```diff
@@ -116,38 +191,55 @@ steps:
       -e DRY_RUN=${DRY_RUN} \
       -e IMAGE_NAME=${IMAGE_NAME} \
       -e VNET_RESOURCE_GROUP_NAME=${VNET_RESOURCE_GROUP_NAME} \
+      -e WINDOWS_SKU=$(WINDOWS_SKU) \
       -e IMPORTED_IMAGE_NAME=${IMPORTED_IMAGE_NAME} \
       -e SIG_GALLERY_NAME=${SIG_GALLERY_NAME} \
+      -e SIG_IMAGE_NAME=${SIG_IMAGE_NAME} \
+      -e GEN2_SIG_FOR_PRODUCTION=${GEN2_SIG_FOR_PRODUCTION} \
       ${CONTAINER_IMAGE} make -f packer.mk cleanup
     displayName: Clean Up Packer Generated Resources
     condition: always()
+
+# Set VHD_NAME and SKU_NAME for Gen 2, which will be published.
+# Note: use -a to grep OS_DISK_SAS (packer-output should be read as a binary file in Linux)
+# Perform this step only if we want to publish the VHD: Default Gen 1 or Gen 2 (sigMode) and the built sig is for production.
   - bash: |
       echo MODE=$(MODE) && \
-      OS_DISK_SAS="$(cat packer-output | grep -a "OSDiskUriReadOnlySas:" | cut -d " " -f 2)" && \
-      VHD_NAME="$(echo $OS_DISK_SAS | cut -d "/" -f 8 | cut -d "?" -f 1)" && \
+      if [[ "${GEN2_SIG_FOR_PRODUCTION}" == "True" ]]; then gen2_captured_sig_version="$(cat vhdbuilder/packer/settings.json | grep "gen2_captured_sig_version" | awk -F':' '{print $2}' | awk -F'"' '{print $2}')"; \
+      [ -n "${gen2_captured_sig_version}" ] && VHD_NAME="${gen2_captured_sig_version}.vhd"; \
+      SKU_NAME="windows-$WINDOWS_SKU"; \
+      else \
+      OS_DISK_SAS="$(cat packer-output | grep -a "OSDiskUriReadOnlySas:" | cut -d " " -f 2)"; \
+      VHD_NAME="$(echo $OS_DISK_SAS | cut -d "/" -f 8 | cut -d "?" -f 1)"; \
+      SKU_NAME="windows-$WINDOWS_SKU"; fi && \
       WINDOWS_VERSION="$(grep -a "OS Version" < release-notes.txt | cut -d ":" -f 2 | tr -d "[:space:]")" && \
       IMAGE_VERSION="$WINDOWS_VERSION.${BUILD_DATE}" && \
-      SKU_NAME="windows-$WINDOWS_SKU" && \
       docker run --rm \
       -v ${PWD}:/go/src/github.com/Azure/AgentBaker \
       -w /go/src/github.com/Azure/AgentBaker \
-      -e CLIENT_ID=${CLIENT_ID} \
-      -e CLIENT_SECRET="$(CLIENT_SECRET)" \
-      -e TENANT_ID=${TENANT_ID} \
-      -e SUBSCRIPTION_ID="${SUBSCRIPTION_ID}" \
+      -e CLIENT_ID=${BUILD_CLIENT_ID} \
```

Comment:
why do we change CLIENT_ID to BUILD_CLIENT_ID?

Reply:
We need to generate sas token for the published vhd but not for the built vhd

Reply:
Fixed.

Reply:
Just to clarify, for the step `build`, `test`, and `cleanup` we are using `BUILD_CLIENT_ID`. For the steps related to publishing `az-copy`, `convert`, and `generate-sas` we are using `CLIENT_ID`?

**Context**: For windows build pipeline `.vsts-vhd-builder-windows.yaml` it has variable `BUILD_CLIENT_ID`. For windows publish pipeline `.vsts-vhd-builder-release-windows.yaml` it has both `BUILD_CLIENT_ID` and `CLIENT_ID`. Both files will reuse the step template which is this file `.builder-release-template-windows.yaml`

For Linux, both build and publish pipeline have only the variable `CLIENT_ID`.

**Question**: I wonder what is the difference.

Reply:
You can check the code which is used to generate sas token. Different credentials are used for different tenant/sub/storage accounts.